tests/pamiq_core/torch/test_trainer.py

```diff
@@ -0,0 +1,159 @@
+import copy
+import logging
+from pathlib import Path
+from typing import Any, override
+
+import pytest
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from pytest_mock import MockerFixture
+
+from pamiq_core.model import InferenceModel, TrainingModel
+from pamiq_core.torch import (
+    LRSchedulersDict,
+    OptimizersDict,
+    OptimizersSetup,
+    StateDict,
+    TorchInferenceModel,
+    TorchTrainer,
+    TorchTrainingModel,
+)
+
+from .are_dict_values_same_entities import are_dict_values_same_entities
+
+logger = logging.getLogger(__name__)
+
+CPU_DEVICE = torch.device("cpu")
+CUDA_DEVICE = torch.device("cuda:0")
+
+
+def get_available_devices() -> list[torch.device]:
+    devices = [CPU_DEVICE]
+    if torch.cuda.is_available():
+        devices.append(CUDA_DEVICE)
+    return devices
+
+
+logger.info("Available devices: " + ", ".join(map(str, get_available_devices())))
+
+parametrize_device = pytest.mark.parametrize("device", get_available_devices())
+
+
+class TorchTrainerImpl(TorchTrainer):
+    @override
+    def create_optimizers(self) -> OptimizersSetup:
+        optimizer_1 = optim.AdamW(
+            self.get_training_model("model_1").model.parameters(),
+            lr=0.001,
+            betas=(0.8, 0.99),
+            weight_decay=0.01,
+        )
+        scheduler_1 = optim.lr_scheduler.ExponentialLR(optimizer_1, gamma=0.998)
+        return {"optimizer_1": optimizer_1}, {"scheduler_1": scheduler_1}
+
+    @override
+    def train(self) -> None:
+        model = self.get_training_model("model_1")
+        device = next(model.model.parameters()).device
+        out = model(torch.randn(5, 2).to(device))
+        self._optimizers["optimizer_1"].zero_grad()
+        out.mean().backward()
+        self._optimizers["optimizer_1"].step()
+        self._schedulers["scheduler_1"].step()
```

Comment:
いくつか想定される書き方があるので、それに従って欲しい。あとデバイスの取得は `get_device`を使う。

```suggestion
class TorchTrainerImpl(TorchTrainer):

    @override
    def on_training_models_attached(self) -> None:
        super().on_training_models_attached()
        self.model_1 = self.get_torch_training_model("model_1", nn.Linear)

    @override
    def create_optimizers(self) -> OptimizersSetup:
        self.optimizer_1 = optim.AdamW(
            self.model_1.model.parameters(),
            lr=0.001,
            betas=(0.8, 0.99),
            weight_decay=0.01,
        )
        self.scheduler_1 = optim.lr_scheduler.ExponentialLR(optimizer_1, gamma=0.998)
        return {"optimizer_1": optimizer_1}, {"scheduler_1": scheduler_1}

    @override
    def train(self) -> None:
        model = self.model_1.model
        device = get_device(model)
        out = model(torch.randn(5, 2).to(device))
        self.optimizer_1.zero_grad()
        out.mean().backward()
        self.optimizer_1.step()
        self.scheduler_1.step()
```